### PR TITLE
expr: pushdown `repeat` to tiflash

### LINKED ISSUE #5114: Implement `Repeat` function push down
## Feature Request

**Describe the feature you'd like:**

We'd like to support `Repeat` function push down.

**Teachability, Documentation, Adoption, Migration Strategy:**

At least one integration test and several unit tests.

### PR DESCRIPTION
**Pull request by likzn**
### What problem does this PR solve?

Issue Number: close #5114 
ref #https://github.com/pingcap/tidb/pull/36024

Problem Summary: push down `repeat` to tiflash

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Support to pushdown repeat to TiFlash
```

### CODE REVIEW

**Comment by likzn** [comment]
Find a bug(maybe)? if we only output one column and there has empty string, it will be truncated. Then I add other columns it will success.
image

**Comment by zanmato1984** [comment]
@likzn Hi, thanks for contribution, we'll look into your PR soon.

Would you please link the corresponding PR in TiDB repo in the description of this one.

**Comment by likzn** [comment]
> @likzn Hi, thanks for contribution, we'll look into your PR soon.
> 
> Would you please link the corresponding PR in TiDB repo in the description of this one?

Done. https://github.com/pingcap/tidb/pull/36024

File: dbms/src/Functions/tests/gtest_strings_repeat.cpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            funcName,
+            toConstString("a"),
+            toConstInt(5)));
+    /// const vec
+    ASSERT_COLUMN_EQ(
+        toVecString({"a", "aa", "aaa", "aaaa"}),
+        executeFunction(
+            funcName,
+            toConstString("a"),
+            toVecInt({1, 2, 3, 4})));
```
Comment (SeaRise) [inline dbms/src/Functions/tests/gtest_strings_repeat.cpp:92]: we should add more cases, such as
- tests for U/Int8, ...U/Int64
- test for Max<IntType>, Min<IntType>
- add some cases like '', '一two三', "\n", etc

Comment (SeaRise) [reply dbms/src/Functions/tests/gtest_strings_repeat.cpp:92]: And test for Null input

File: dbms/src/Functions/FunctionsString.cpp [hunk 2]
```diff
@@ -4188,6 +4188,212 @@ class FunctionLength : public IFunction
...
+            constVector(col_string_from_const->getChars(),
+                        col_string_from_const->getOffsets(),
+                        col_vector_repeat_count->getData(),
+                        col_res->getChars(),
+                        col_res->getOffsets());
+        }
+        else
+        {
+            // Impossible to reach here
+            return false;
```
Comment (guo-shaoge) [inline dbms/src/Functions/FunctionsString.cpp:4300]: Maybe throw exception?

File: dbms/src/Functions/FunctionsString.cpp [hunk 3]
```diff
@@ -4188,6 +4188,212 @@ class FunctionLength : public IFunction
...
+        const ColumnString::Offsets & offsets,
+        const PaddedPODArray<IntType> & repeat_times,
+        ColumnString::Chars_t & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        size_t size = repeat_times.size();
+        res_offsets.resize(size);
+
+        ColumnString::Offset start_offset = 0;
+        ColumnString::Offset end_offset = offsets[0];
```
Comment (guo-shaoge) [inline dbms/src/Functions/FunctionsString.cpp:4360]:
```suggestion
        const ColumnString::Offset start_offset = 0;
        const ColumnString::Offset end_offset = offsets[0];
```

File: tests/tidb-ci/fullstack-test-dt/repeat_push_down.test [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+mysql> drop table if exists test.f;
+mysql> CREATE TABLE test.f (a char(20),b int, id int);
+mysql> insert into test.f values('abc',1,1);
+mysql> insert into test.f values('pingcap',2,1);
+mysql> insert into test.f values('li',-1,1);
+mysql> insert into test.f values(null,4,1);
+mysql> insert into test.f values('ww',null,1);
+mysql> alter table test.f set tiflash replica 1;
+
```
[the remark this answers is not in the document]
Comment (guo-shaoge) [reply tests/tidb-ci/fullstack-test-dt/repeat_push_down.test:24]: Rest LGTM

File: tests/tidb-ci/fullstack-test-dt/repeat_push_down.test [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+mysql> use test; set @@tidb_allow_mpp=1;set @@tidb_enforce_mpp=1;  select a,b,repeat(a,b) from test.f;
```
Comment (guo-shaoge) [inline tests/tidb-ci/fullstack-test-dt/repeat_push_down.test:25]:
```suggestion
mysql> use test; set @@tidb_isolation_read_engines='tiflash'; set @@tidb_allow_mpp=1;set @@tidb_enforce_mpp=1;  select a,b,repeat(a,b) from test.f;
```

Better to add `tidb_isolation_read_engines`. Will give error when tiflash replica is not ready.
When only tidb_enforce_mpp is set, will give **warning** instead.

File: dbms/src/Functions/FunctionsString.cpp [hunk 6]
```diff
@@ -4188,6 +4188,212 @@ class FunctionLength : public IFunction
...
+            {
+                const ColumnConst * col_const_repeat_times = checkAndGetColumnConst<ColumnVector<IntType>>(column_repeat_times.get());
+                if (col_const_repeat_times == nullptr)
+                {
+                    return false;
+                }
+                auto repeat_times = col_const_repeat_times->getValue<IntType>();
+                vectorConst(col_string->getChars(),
+                            col_string->getOffsets(),
+                            accurate::lessOp(INT64_MAX, repeat_times) ? INT64_MAX : repeat_times,
```
Comment (guo-shaoge) [inline dbms/src/Functions/FunctionsString.cpp:4263]: Why use `INT64_MAX`, looks like tidb use `INT32_MAX`. Check https://github.com/pingcap/tidb/blob/fb6f6a4fda0fa24ec1f5ba3f1d31225c7fc4814a/expression/builtin_string.go#L675.

Maybe better to keep it compatible.

Comment (likzn) [reply dbms/src/Functions/FunctionsString.cpp]: In Tiflash, how did we deal with `maxAllowedPacket`. I can not find similar code
`if uint64(byteLength)*uint64(num) > b.maxAllowedPacket {
		return "", true, handleAllowedPacketOverflowed(b.ctx, "repeat", b.maxAllowedPacket)
	}`

Comment (guo-shaoge) [reply dbms/src/Functions/FunctionsString.cpp]: I think no need to handle it for now. Please add a comment in this issue as a remainder. https://github.com/pingcap/tiflash/issues/3669

**Comment by likzn** [comment]
TODO: We should handle `maxAllowedPacket` about String function together. #3669

File: dbms/src/Functions/FunctionsString.cpp (hunk 6 above)
Comment (likzn) [reply dbms/src/Functions/FunctionsString.cpp]: Done. Also add `todo` in code.

**Comment by SeaRise** [comment]
Add release note like 'Support to pushdown repeat to TiFlash'.
And add the pr link of tidb?

**Comment by likzn** [comment]
> Add release note like 'Support to pushdown repeat to TiFlash'. And add the pr link of tidb?

Done.

**Review by SeaRise — APPROVED** [review]

**Review by guo-shaoge — APPROVED** [review]

**Pull request merged**